File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/datatablerowdialog/DataTableRowDialogFragment.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */

package com.mifos.mifosxdroid.dialogfragments.datatablerowdialog;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.Toast;

import androidx.fragment.app.DialogFragment;

import com.mifos.api.GenericResponse;
import com.mifos.exceptions.RequiredFieldException;
import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.core.MifosBaseActivity;
import com.mifos.mifosxdroid.databinding.DialogFragmentAddEntryToDatatableBinding;
import com.mifos.mifosxdroid.formwidgets.FormEditText;
import com.mifos.mifosxdroid.formwidgets.FormNumericEditText;
import com.mifos.mifosxdroid.formwidgets.FormSpinner;
import com.mifos.mifosxdroid.formwidgets.FormToggleButton;
import com.mifos.mifosxdroid.formwidgets.FormWidget;
import com.mifos.objects.noncore.ColumnHeader;
import com.mifos.objects.noncore.ColumnValue;
import com.mifos.objects.noncore.DataTable;
import com.mifos.utils.Constants;
import com.mifos.utils.SafeUIBlockingUtility;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;


/**
 * Created by ishankhanna on 01/08/14.
 */
public class DataTableRowDialogFragment extends DialogFragment
        implements DataTableRowDialogMvpView {

    private final String LOG_TAG = getClass().getSimpleName();
    private DialogFragmentAddEntryToDatatableBinding binding;


    @Inject
    DataTableRowDialogPresenter dataTableRowDialogPresenter;

    private DataTable dataTable;
    private int entityId;
    private SafeUIBlockingUtility safeUIBlockingUtility;
    private List<FormWidget> listFormWidgets = new ArrayList<>();


    //TODO Check for Static vs Bundle Approach
    public static DataTableRowDialogFragment newInstance(DataTable dataTable, int entityId) {
        DataTableRowDialogFragment dataTableRowDialogFragment = new DataTableRowDialogFragment();
        Bundle args = new Bundle();
        dataTableRowDialogFragment.dataTable = dataTable;
        dataTableRowDialogFragment.entityId = entityId;
        dataTableRowDialogFragment.setArguments(args);
        return dataTableRowDialogFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((MifosBaseActivity) getActivity()).getActivityComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle
            savedInstanceState) {

        /**
         * This is very Important
         * It is used to auto resize the dialog when a Keyboard appears.
         * And User can still easily scroll through the form. Sweet, isn't it?
         */
        getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams
                .SOFT_INPUT_ADJUST_RESIZE);

        binding = DialogFragmentAddEntryToDatatableBinding.inflate(inflater, container, false);
        dataTableRowDialogPresenter.attachView(this);


        getDialog().setTitle(dataTable.getRegisteredTableName());

        safeUIBlockingUtility = new SafeUIBlockingUtility(DataTableRowDialogFragment.this
                .getActivity(), getString(R.string.data_table_row_dialog_loading_message));

        createForm(dataTable);
        addSaveButton();

        return binding.getRoot();
    }

    public void createForm(DataTable table) {
        List<FormWidget> formWidgets = new ArrayList<>();

        for (ColumnHeader columnHeader : table.getColumnHeaderData()) {

            if (!columnHeader.getColumnPrimaryKey()) {

                if (columnHeader.getColumnDisplayType().equals(FormWidget.SCHEMA_KEY_STRING) ||
                        columnHeader.getColumnDisplayType().equals(FormWidget.SCHEMA_KEY_TEXT)) {

                    FormEditText formEditText = new FormEditText(getActivity(), columnHeader
                            .getColumnName());
                    formWidgets.add(formEditText);
                    binding.llDataTableEntryForm.addView(formEditText.getView());

                } else if (columnHeader.getColumnDisplayType().equals(FormWidget.SCHEMA_KEY_INT)) {

                    FormNumericEditText formNumericEditText = new FormNumericEditText(getActivity
                            (), columnHeader.getColumnName());
                    formNumericEditText.setReturnType(FormWidget.SCHEMA_KEY_INT);
                    formWidgets.add(formNumericEditText);
                    binding.llDataTableEntryForm.addView(formNumericEditText.getView());


                } else if (columnHeader.getColumnDisplayType().equals(FormWidget
                        .SCHEMA_KEY_DECIMAL)) {

                    FormNumericEditText formNumericEditText = new FormNumericEditText(getActivity
                            (), columnHeader.getColumnName());
                    formNumericEditText.setReturnType(FormWidget.SCHEMA_KEY_DECIMAL);
                    formWidgets.add(formNumericEditText);
                    binding.llDataTableEntryForm.addView(formNumericEditText.getView());


                } else if (columnHeader.getColumnDisplayType().equals(FormWidget
                        .SCHEMA_KEY_CODELOOKUP) || columnHeader.getColumnDisplayType().equals
                        (FormWidget.SCHEMA_KEY_CODEVALUE)) {

                    if (columnHeader.getColumnValues().size() > 0) {
                        List<String> columnValueStrings = new ArrayList<>();
                        List<Integer> columnValueIds = new ArrayList<>();

                        for (ColumnValue columnValue : columnHeader.getColumnValues()) {
                            columnValueStrings.add(columnValue.getValue());
                            columnValueIds.add(columnValue.getId());
                        }

                        FormSpinner formSpinner = new FormSpinner(getActivity(), columnHeader
                                .getColumnName(), columnValueStrings, columnValueIds);
                        formSpinner.setReturnType(FormWidget.SCHEMA_KEY_CODEVALUE);
                        formWidgets.add(formSpinner);
                        binding.llDataTableEntryForm.addView(formSpinner.getView());
                    }

                } else if (columnHeader.getColumnDisplayType().equals(FormWidget.SCHEMA_KEY_DATE)) {

                    FormEditText formEditText = new FormEditText(getActivity(), columnHeader
                            .getColumnName());
                    formEditText.setIsDateField(true, getActivity().getSupportFragmentManager());
                    formWidgets.add(formEditText);
                    binding.llDataTableEntryForm.addView(formEditText.getView());
                } else if (columnHeader.getColumnDisplayType().equals(FormWidget.SCHEMA_KEY_BOOL)) {

                    FormToggleButton formToggleButton = new FormToggleButton(getActivity(),
                            columnHeader.getColumnName());
                    formWidgets.add(formToggleButton);
                    binding.llDataTableEntryForm.addView(formToggleButton.getView());
                }
            }
        }
        listFormWidgets.addAll(formWidgets);
    }

    private void addSaveButton() {
        Button bt_processForm = new Button(getActivity());
        bt_processForm.setLayoutParams(FormWidget.defaultLayoutParams);
        bt_processForm.setText(getString(R.string.save));
        bt_processForm.setBackgroundColor(getActivity().getResources().getColor(R.color.blue_dark));

        binding.llDataTableEntryForm.addView(bt_processForm);
        bt_processForm.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    onSaveActionRequested();
                } catch (RequiredFieldException e) {
                    Log.d(LOG_TAG, e.getMessage());
                }
            }
        });
    }

    public void onSaveActionRequested() throws RequiredFieldException {
        dataTableRowDialogPresenter.addDataTableEntry(dataTable.getRegisteredTableName(),
                entityId, addDataTableInput());
    }

    private HashMap<String, Object> addDataTableInput() {
        List<FormWidget> formWidgets = listFormWidgets;
        HashMap<String, Object> payload = new HashMap<>();
        payload.put(Constants.DATE_FORMAT, "dd-mm-YYYY");
        payload.put(Constants.LOCALE, "en");
        for (FormWidget formWidget : formWidgets) {
            if (formWidget.getReturnType().equals(FormWidget.SCHEMA_KEY_INT)) {
                payload.put(formWidget.getPropertyName(), Integer.parseInt(formWidget.getValue()
                        .equals("") ? "0" : formWidget.getValue()));
            } else if (formWidget.getReturnType().equals(FormWidget.SCHEMA_KEY_DECIMAL)) {
                payload.put(formWidget.getPropertyName(), Double.parseDouble(formWidget.getValue
                        ().equals("") ? "0.0" : formWidget.getValue()));
            } else if (formWidget.getReturnType().equals(FormWidget.SCHEMA_KEY_CODEVALUE)) {
                FormSpinner formSpinner = (FormSpinner) formWidget;
                payload.put(formWidget.getPropertyName(), formSpinner.getIdOfSelectedItem
                        (formWidget.getValue()));
            } else {
                payload.put(formWidget.getPropertyName(), formWidget.getValue());
            }
        }
        return payload;
    }

    @Override
    public void showDataTableEntrySuccessfully(GenericResponse genericResponse) {
        Toast.makeText(getActivity(), R.string.data_table_entry_added, Toast.LENGTH_LONG).show();
        getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK,
                getActivity().getIntent());
        getActivity().getSupportFragmentManager().popBackStack();
    }

    @Override
    public void showError(String message) {
        Toast.makeText(getActivity(), message, Toast.LENGTH_LONG).show();
        getActivity().getSupportFragmentManager().popBackStack();
    }

    @Override
    public void showProgressbar(boolean b) {
        if (b) {
            safeUIBlockingUtility.safelyBlockUI();
        } else {
            safeUIBlockingUtility.safelyUnBlockUI();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        dataTableRowDialogPresenter.detachView();
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/datatablerowdialog/DataTableRowDialogFragment.kt
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.mifosxdroid.dialogfragments.datatablerowdialog

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Button
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.mifos.core.network.GenericResponse
import com.mifos.core.objects.noncore.DataTable
import com.mifos.exceptions.RequiredFieldException
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.databinding.DialogFragmentAddEntryToDatatableBinding
import com.mifos.mifosxdroid.formwidgets.FormEditText
import com.mifos.mifosxdroid.formwidgets.FormNumericEditText
import com.mifos.mifosxdroid.formwidgets.FormSpinner
import com.mifos.mifosxdroid.formwidgets.FormToggleButton
import com.mifos.mifosxdroid.formwidgets.FormWidget
import com.mifos.utils.Constants
import com.mifos.utils.SafeUIBlockingUtility
import dagger.hilt.android.AndroidEntryPoint

/**
 * Created by ishankhanna on 01/08/14.
 */
@AndroidEntryPoint
class DataTableRowDialogFragment : DialogFragment() {

    private lateinit var binding: DialogFragmentAddEntryToDatatableBinding

    private val LOG_TAG = javaClass.simpleName

    private lateinit var viewModel: DataTableRowDialogViewModel

    private var dataTable: DataTable? = null
    private var entityId = 0
    private var safeUIBlockingUtility: SafeUIBlockingUtility? = null
    private val listFormWidgets: MutableList<FormWidget> = ArrayList()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        /**
         * This is very Important
         * It is used to auto resize the dialog when a Keyboard appears.
         * And User can still easily scroll through the form. Sweet, isn't it?
         */
        dialog?.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
        binding = DialogFragmentAddEntryToDatatableBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this)[DataTableRowDialogViewModel::class.java]
        dialog?.setTitle(dataTable?.registeredTableName)
        safeUIBlockingUtility = SafeUIBlockingUtility(
            requireContext(), getString(R.string.data_table_row_dialog_loading_message)
        )
        createForm(dataTable)
        addSaveButton()

        viewModel.dataTableRowDialogUiState.observe(viewLifecycleOwner) {
            when (it) {
                is DataTableRowDialogUiState.ShowDataTableEntrySuccessfully -> {
                    showProgressbar(false)
                    showDataTableEntrySuccessfully(it.genericResponse)
                }

                is DataTableRowDialogUiState.ShowError -> {
                    showProgressbar(false)
                    showError(it.message)
                }

                is DataTableRowDialogUiState.ShowProgressbar -> showProgressbar(true)
            }
        }

        return binding.root
    }

    private fun createForm(table: DataTable?) {
        val formWidgets: MutableList<FormWidget> = ArrayList()
        if (table != null) {
            for (columnHeader in table.columnHeaderData) {
                if (!columnHeader.columnPrimaryKey!!) {
                    if (columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_STRING || columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_TEXT) {
                        val formEditText = FormEditText(
                            activity, columnHeader
                                .dataTableColumnName
                        )
                        formWidgets.add(formEditText)
                        binding.llDataTableEntryForm.addView(formEditText.view)
                    } else if (columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_INT) {
                        val formNumericEditText =
                            FormNumericEditText(activity, columnHeader.dataTableColumnName)
                        formNumericEditText.returnType = FormWidget.SCHEMA_KEY_INT
                        formWidgets.add(formNumericEditText)
                        binding.llDataTableEntryForm.addView(formNumericEditText.view)
                    } else if (columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_DECIMAL) {
                        val formNumericEditText =
                            FormNumericEditText(activity, columnHeader.dataTableColumnName)
                        formNumericEditText.returnType = FormWidget.SCHEMA_KEY_DECIMAL
                        formWidgets.add(formNumericEditText)
                        binding.llDataTableEntryForm.addView(formNumericEditText.view)
                    } else if (columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_CODELOOKUP || columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_CODEVALUE) {
                        if (columnHeader.columnValues.isNotEmpty()) {
                            val columnValueStrings: MutableList<String> = ArrayList()
                            val columnValueIds: MutableList<Int> = ArrayList()
                            for (columnValue in columnHeader.columnValues) {
                                columnValue.value?.let { columnValueStrings.add(it) }
                                columnValue.id?.let { columnValueIds.add(it) }
                            }
                            val formSpinner = FormSpinner(
                                activity, columnHeader
                                    .dataTableColumnName, columnValueStrings, columnValueIds
                            )
                            formSpinner.returnType = FormWidget.SCHEMA_KEY_CODEVALUE
                            formWidgets.add(formSpinner)
                            binding.llDataTableEntryForm.addView(formSpinner.view)
                        }
                    } else if (columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_DATE) {
                        val formEditText = FormEditText(
                            activity, columnHeader
                                .dataTableColumnName
                        )
                        formEditText.setIsDateField(true, requireActivity().supportFragmentManager)
                        formWidgets.add(formEditText)
                        binding.llDataTableEntryForm.addView(formEditText.view)
                    } else if (columnHeader.columnDisplayType == FormWidget.SCHEMA_KEY_BOOL) {
                        val formToggleButton = FormToggleButton(
                            activity,
                            columnHeader.dataTableColumnName
                        )
                        formWidgets.add(formToggleButton)
                        binding.llDataTableEntryForm.addView(formToggleButton.view)
                    }
                }
            }
        }
        listFormWidgets.addAll(formWidgets)
    }

    private fun addSaveButton() {
        val bt_processForm = Button(activity)
        bt_processForm.layoutParams = FormWidget.defaultLayoutParams
        bt_processForm.text = getString(R.string.save)
        bt_processForm.setBackgroundColor(requireActivity().resources.getColor(R.color.blue_dark))
        binding.llDataTableEntryForm.addView(bt_processForm)
        bt_processForm.setOnClickListener {
            try {
                onSaveActionRequested()
            } catch (e: RequiredFieldException) {
                Log.d(LOG_TAG, e.message.toString())
            }
        }
    }

    private fun onSaveActionRequested() {
        viewModel.addDataTableEntry(
            dataTable?.registeredTableName,
            entityId, addDataTableInput()
        )
    }

    private fun addDataTableInput(): HashMap<String, String> {
        val formWidgets: List<FormWidget> = listFormWidgets
        val payload = HashMap<String, String>()
        payload[Constants.DATE_FORMAT] = "dd-mm-YYYY"
        payload[Constants.LOCALE] = "en"
        for (formWidget in formWidgets) {
            when (formWidget.returnType) {
                FormWidget.SCHEMA_KEY_INT -> payload[formWidget.propertyName] =
                    (if (formWidget.value
                        == ""
                    ) "0" else formWidget.value).toInt().toString()

                FormWidget.SCHEMA_KEY_DECIMAL -> payload[formWidget.propertyName] =
                    (if (formWidget.value == "") "0.0" else formWidget.value).toDouble().toString()

                FormWidget.SCHEMA_KEY_CODEVALUE -> {
                    val formSpinner = formWidget as FormSpinner
                    payload[formWidget.propertyName] =
                        formSpinner.getIdOfSelectedItem(formWidget.value).toString()
                }

                else -> payload[formWidget.propertyName] = formWidget.value
            }
        }
        return payload
    }

    private fun showDataTableEntrySuccessfully(genericResponse: GenericResponse) {
        Toast.makeText(activity, R.string.data_table_entry_added, Toast.LENGTH_LONG).show()
        targetFragment?.onActivityResult(
            targetRequestCode, Activity.RESULT_OK,
            requireActivity().intent
        )
        requireActivity().supportFragmentManager.popBackStack()
    }

    private fun showError(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_LONG).show()
        requireActivity().supportFragmentManager.popBackStack()
    }

    private fun showProgressbar(b: Boolean) {
        if (b) {
            safeUIBlockingUtility?.safelyBlockUI()
        } else {
            safeUIBlockingUtility?.safelyUnBlockUI()
        }
    }

    companion object {
        //TODO Check for Static vs Bundle Approach
        fun newInstance(dataTable: DataTable?, entityId: Int): DataTableRowDialogFragment {
            val dataTableRowDialogFragment = DataTableRowDialogFragment()
            val args = Bundle()
            dataTableRowDialogFragment.dataTable = dataTable
            dataTableRowDialogFragment.entityId = entityId
            dataTableRowDialogFragment.arguments = args
            return dataTableRowDialogFragment
        }
    }
}